ros_ws/src/autonomous/src/wall_following.cpp

```diff
@@ -0,0 +1,336 @@
+
+	ros::Subscriber lidar_subscriber;	
+ 	void lidarCallback(const sensor_msgs::LaserScan::ConstPtr& lidar);
+
+	ros::Publisher pid_publisher;
+
+	
+	// void adjustSpeed(double speed);
+};
+
+WallFollowing::WallFollowing()
+{
+	lidar_subscriber = nh_.subscribe<sensor_msgs::LaserScan>("/racer/laser/scan", 1, &WallFollowing::lidarCallback, this);
+	pid_publisher = nh_.advertise< drive_msgs::pid_input >("/pid_input", 1);
+}
+
+/**
+ * @brief Using the lidar scans and a given angle theta
+ * this method returns the range at the given theta after doing boundary checks.
+ * 
+ * @param lidar 
+ * @param theta 
+ * @return float 
+ */
+float getRange(const sensor_msgs::LaserScan::ConstPtr& lidar, float theta)
+{	
+	// ranges size is 720, angle range is 270°
+	// so one step equals an angle of 0,375°
+	// to access range at angle theta, you have to divide the given theta by the step size
+	float range = lidar->ranges[theta/0.375];
+
+	// check for absurd or nan values
+	if(range > 4 || std::isinf(range))
+	{
+		range = 4;
+	}
+	return range;
+}
+
+/**
+ * @brief Checks if there is an unavoidable wall in front of the car using the lidar scans
+ * and returns a signal (0) to inhibit driving if it is the case.
+ * 
+ * @param lidar 
+ * @return float 
+ */
+float emergencyStop(const sensor_msgs::LaserScan::ConstPtr& lidar)
+{
+
+	// used for range averaging
+	float front_range_sum = 0;
+
+	// calculate the average distance in front of the car (6° radius)
+	// used for robustness 
+	// (e.g. there is noise and only a single index shows a close range...
+	// ... this is probably not an obstacle)
+	for(int i = 133; i < 139; i++)
+	{
+		front_range_sum += getRange(lidar, i);
+		// ROS_INFO_STREAM("front range sum: " << front_range_sum);
+	}
+
+	// return 0 (stop) if the object is too close
+	if((front_range_sum/6) < 0.3)
+	{
+		ROS_INFO_STREAM("too close!: " << front_range_sum);
+		return 0;
+	}
+	else
+	{
+		return 1;
+	}
+}
+
+/**
+ * @brief Using the lidar scans this method calculates the necesarry angle and velocity
+ * for the car to stay parallel to the right wall and then returns them both in that order.
+ * 
+ * @param lidar 
+ * @return std::array<float, 2> 
+ */
+
+std::array<float, 2> followRightWall(const sensor_msgs::LaserScan::ConstPtr& lidar)
+{
+	// calculations done according to the F1Tenth tutorial
+	// http://f1tenth.org/lab_instructions/t6.pdf
+
+	// some angle theta between 0 and 70 which you can choose by yourself
+	float theta = 95;
+
+	// two distances a and b
+	// with b being the distance at angle 0 and a being the distance at angle theta
+	// the lidar has a range of 270°, the algorithm use 180°
+	// so the the angle 0° for the algorith starts at angle 45° for the lidar
+	float a = getRange(lidar,theta);
+	float b = getRange(lidar,45);
+
+	ROS_INFO_STREAM("right a: " << a);
+	ROS_INFO_STREAM("right b: " << b);
+
+	// transform the difference between the angles at a and b into radians
+	float swing = (fabs(theta - 45) * M_PI) / 180;
```

Comment:
May replace the call of `fabs` with the namespaced version `std::fabs` or `std::abs` with the later having the advantage of not (even though in this case rightly) giving anyone the assumption the number given or returned has to be of type `float`.